// @datalayer/jupyter-react/lib/components/notebook/cell/sidebar/CellSidebarWidget.js
import { jsx as _jsx } from "react/jsx-runtime";
/*
 * Copyright (c) 2021-2023 Datalayer, Inc.
 *
 * MIT License
 */
import { createElement } from 'react';
import { createPortal } from 'react-dom';
import { newUuid } from '../../../../utils/Utils';
import { ReactPortalWidget } from '../../../lumino/ReactPortalWidget';
import { notebookActions } from '../../NotebookRedux';
export const DATALAYER_CELL_HEADER_CLASS = 'dla-CellHeader-Container';
export class CellSidebarWidget extends ReactPortalWidget {
    commands;
    constructor(CellSidebar, notebookId, nbgrader, commands, store) {
        super();
        this.commands = commands;
        this.addClass('jp-CellHeader');
        this.id = newUuid();
        const props = {
            notebookId: notebookId,
            cellId: this.id,
            command: this.commands,
            nbgrader,
        };
        const sidebar = createElement(CellSidebar, props);
        const portalDiv = (_jsx("div", { className: DATALAYER_CELL_HEADER_CLASS, children: sidebar }));
        const portal = createPortal(portalDiv, this.node);
        store.dispatch(notebookActions.addPortals({
            uid: notebookId,
            portals: [portal],
        }));
    }
}
export default CellSidebarWidget;
