// @datalayer/primer-addons/lib/components/content-loader/stories/ContentLoader.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ThemeProvider, BaseStyles, Box } from "@primer/react";
import { ContentLoader } from '../../../index';
const meta = {
    title: 'Components/ContentLoader',
    component: ContentLoader,
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs'],
    parameters: {
        // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
        layout: 'fullscreen',
    },
};
export default meta;
const ThemedContentLoader = (props) => _jsx(ThemeProvider, { colorMode: props.colorMode, children: _jsx(BaseStyles, { children: _jsx(Box, { p: 3, bg: "canvas.default", children: _jsx(ContentLoader, { ...props }) }) }) });
export const ContentLoaderDay = {
    args: {
        count: 3
    },
    render: (args) => _jsx(ThemedContentLoader, { ...args, colorMode: "day" })
};
export const ContentLoaderNight = {
    args: {
        count: 3
    },
    render: (args) => _jsx(ThemedContentLoader, { ...args, colorMode: "night" })
};
