// @primer/react/lib/drafts/Skeleton/SkeletonAvatar.js
'use strict';

var React = require('react');
var getBreakpointDeclarations = require('../../utils/getBreakpointDeclarations.js');
var constants = require('../../constants.js');
var useResponsiveValue = require('../../hooks/useResponsiveValue.js');
var Avatar = require('../../Avatar/Avatar.js');
var SkeletonBox = require('./SkeletonBox.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const avatarSkeletonStyles = {
  '&[data-component="SkeletonAvatar"]': {
    borderRadius: '50%',
    boxShadow: `0 0 0 1px ${constants.get('colors.avatar.border')}`,
    display: 'inline-block',
    lineHeight: constants.get('lineHeights.condensedUltra'),
    height: 'var(--avatar-size)',
    width: 'var(--avatar-size)'
  },
  '&[data-avatar-shape="square"]': {
    borderRadius: 'clamp(4px, var(--avatar-size) - 24px, 6px)'
  }
};
const SkeletonAvatar = ({
  size = Avatar.DEFAULT_AVATAR_SIZE,
  square,
  ...rest
}) => {
  const avatarSx = useResponsiveValue.isResponsiveValue(size) ? {
    ...getBreakpointDeclarations.getBreakpointDeclarations(size, '--avatar-size', value => `${value || Avatar.DEFAULT_AVATAR_SIZE}px`),
    ...avatarSkeletonStyles
  } : {
    '--avatar-size': `${size}px`,
    ...avatarSkeletonStyles
  };
  return /*#__PURE__*/React__default.default.createElement(SkeletonBox.SkeletonBox, _extends({
    sx: avatarSx
  }, rest, {
    "data-component": "SkeletonAvatar",
    "data-avatar-shape": square ? 'square' : undefined
  }));
};
SkeletonAvatar.displayName = "SkeletonAvatar";

exports.SkeletonAvatar = SkeletonAvatar;
