// @datalayer/primer-addons/lib/components/toolbar/stories/Toolbar.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ThemeProvider, BaseStyles, Box, Text, IconButton, } from "@primer/react";
import { SearchIcon, ThreeBarsIcon, GearIcon } from "@primer/octicons-react";
import { Toolbar } from '../../../index';
const meta = {
    title: 'Components/Toolbar',
    component: Toolbar,
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs'],
    parameters: {
        // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
        layout: 'fullscreen',
    },
};
export default meta;
const ThemedToolbar = (props) => _jsx(ThemeProvider, { colorMode: props.colorMode, children: _jsx(BaseStyles, { children: _jsxs(Toolbar, { ...props, heading: _jsxs(Box, { display: "flex", alignItems: "center", sx: { gap: "0.5rem" }, children: [_jsx(IconButton, { "aria-label": "Search", icon: ThreeBarsIcon, size: "small", variant: "invisible" }), _jsx(Text, { children: "Datalayer Toolbar" })] }), children: [_jsx(IconButton, { "aria-label": "Search", icon: SearchIcon, size: "small" }), _jsx(Text, { children: "Action1" }), _jsx(Text, { children: "Action2" }), _jsx(Text, { children: "Action3" }), _jsx(IconButton, { "aria-label": "Search", icon: GearIcon, size: "small", variant: "invisible" })] }) }) });
export const ToolbarDay = {
    args: {
        align: "flex-end",
        border: true,
        gap: "1rem",
        padding: "0 1rem",
        children: [],
    },
    render: (args) => _jsx(ThemedToolbar, { ...args, colorMode: "day" })
};
export const ToolbarNight = {
    args: {
        align: "flex-end",
        border: true,
        gap: "1rem",
        padding: "0 1rem",
        children: [],
    },
    render: (args) => _jsx(ThemedToolbar, { ...args, colorMode: "night" })
};
