// @datalayer/jupyter-react/lib/components/notebook/content/JupyterReactContentFactory.js
/*
 * Copyright (c) 2021-2023 Datalayer, Inc.
 *
 * MIT License
 */
import { NotebookPanel } from '@jupyterlab/notebook';
import { CellSidebarWidget, } from '../cell/sidebar/CellSidebarWidget';
// import { IInputPrompt } from '@jupyterlab/cells';
// import { NotebookInputPrompt } from './../cell/InputPrompt';
/**
 * Extend the default implementation NotebookPanel.ContentFactory of `IContentFactory`.
 */
export class JupyterReactContentFactory extends NotebookPanel.ContentFactory {
    CellSidebar;
    notebookId;
    nbgrader;
    commands;
    store;
    constructor(CellSidebar, notebookId, nbgrader, commands, options, store) {
        super(options);
        this.CellSidebar = CellSidebar;
        super.createCodeCell;
        (this.notebookId = notebookId), (this.nbgrader = nbgrader);
        this.commands = commands;
        this.store = store;
    }
    /** @override */
    createCellHeader() {
        return new CellSidebarWidget(this.CellSidebar, this.notebookId, this.nbgrader, this.commands, this.store);
    }
}
export default JupyterReactContentFactory;
