// @datalayer/primer-addons/lib/components/slider/stories/Slider.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from 'react';
import { ThemeProvider, BaseStyles, Box, CounterLabel } from "@primer/react";
import { Slider } from '../../../index';
const meta = {
    title: 'Components/Slider',
    component: Slider,
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs'],
    parameters: {
        // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
        layout: 'fullscreen',
    },
};
export default meta;
const ThemedSlider = (props) => {
    const [value, setValue] = useState(50);
    return (_jsx(ThemeProvider, { colorMode: props.colorMode, children: _jsx(BaseStyles, { children: _jsxs(Box, { p: 3, bg: "canvas.default", children: [_jsx(Box, { children: _jsx(Slider, { ...props, value: value, onChange: setValue }) }), _jsx(Box, { children: _jsx(CounterLabel, { sx: { ml: 2 }, children: value }) })] }) }) }));
};
export const SliderDay = {
    args: {
        name: "slider-day",
        id: "slider-day",
        min: 0,
        max: 100,
        step: 10,
        value: 50,
        onChange: (value) => {
            console.log('Slider value changed:', value);
        },
    },
    render: (args) => _jsx(ThemedSlider, { ...args, colorMode: "day" })
};
export const SliderNight = {
    args: {
        name: "slider-night",
        id: "slider-night",
        min: 0,
        max: 100,
        step: 10,
        value: 50,
        onChange: (value) => {
            console.log('Slider value changed:', value);
        },
    },
    render: (args) => _jsx(ThemedSlider, { ...args, colorMode: "night" })
};
