// @datalayer/primer-addons/lib/components/toolbar/Toolbar.js
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box } from "@primer/react";
export const Toolbar = ({ align = "flex-start", height = "5vh", gap = "0.1rem", padding, heading, backgroundColor, borderBottomHeight = 1, borderBottomColor, border = false, children, }) => {
    return (_jsxs(Box, { display: "flex", alignItems: "center", sx: {
            backgroundColor: backgroundColor ? backgroundColor : "pageHeaderBg",
            height: height,
            width: "100%",
            padding: padding,
            borderBottomWidth: borderBottomHeight,
            borderBottomStyle: border ? "solid" : "none",
            borderColor: borderBottomColor ? borderBottomColor : "border.default",
        }, children: [_jsx(Box, { display: "flex", children: heading != null && heading != undefined ? heading : _jsx(_Fragment, {}) }), _jsx(Box, { display: "flex", flexGrow: 1, alignItems: "center", justifyContent: align, sx: {
                    gap: gap,
                }, children: children })] }));
};
export default Toolbar;
