// @datalayer/primer-addons/lib/components/closeable-flash/stories/CloseableFlash.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ThemeProvider, BaseStyles, Box } from "@primer/react";
import { CheckIcon } from "@primer/octicons-react";
import { CloseableFlash } from '../../../index';
const meta = {
    title: 'Components/CloseableFlash',
    component: CloseableFlash,
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs'],
    parameters: {
        // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
        layout: 'fullscreen',
    },
};
export default meta;
const ThemedCloseableFlash = (props) => {
    return (_jsx(ThemeProvider, { colorMode: props.colorMode, children: _jsx(BaseStyles, { children: _jsx(Box, { p: 3, bg: "canvas.default", children: _jsx(Box, { children: _jsx(CloseableFlash, { ...props }) }) }) }) }));
};
export const CloseableFlashDay = {
    args: {
        variant: "success",
        leadingIcon: CheckIcon,
        onClose: () => window.alert("closed")
    },
    render: (args) => _jsx(ThemedCloseableFlash, { ...args, colorMode: "day", children: "Success Flash" })
};
export const CloseableFlashNight = {
    args: {
        variant: "success",
        leadingIcon: CheckIcon,
        onClose: () => window.alert("closed")
    },
    render: (args) => _jsx(ThemedCloseableFlash, { ...args, colorMode: "night", children: "Success Flash" })
};
